client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -306,6 +307,34 @@ private long getRemainingBytes(SegmentMetadataClientFactory metaFactory, StreamC
         return cuts;
     }
 
+    @Override
+    public CompletableFuture<Map<Stream, StreamCut>> getCurrentStreamCut(ScheduledExecutorService backgroundExecutor) {
+        String internalChkPointId = UUID.randomUUID().toString();
+        log.debug("Fetching the current StreamCut using id {}", internalChkPointId);
+        synchronizer.updateStateUnconditionally(new CreateCheckpoint(internalChkPointId, true));
+        AtomicBoolean fetchStreamCutPending = new AtomicBoolean(true);
+
+        return Futures.loop(fetchStreamCutPending::get, () -> {
+            return Futures.delayedTask(() -> {
+                synchronizer.fetchUpdates();
+                fetchStreamCutPending.set(!synchronizer.getState().isCheckpointComplete(internalChkPointId));
+                if (fetchStreamCutPending.get()) {
+                    log.debug("Waiting on fetch StreamCut with id {}, currentState is: {}", internalChkPointId, synchronizer.getState());
+                }
+                return null;
+            }, Duration.ofMillis(500), backgroundExecutor);
+        }, backgroundExecutor)
+                      .thenApply(v -> completeChkPointAndFetchStreamCut(internalChkPointId));
+    }
+
+    @SneakyThrows(CheckpointFailedException.class)
+    private Map<Stream, StreamCut> completeChkPointAndFetchStreamCut(String checkPointId) {
+        ReaderGroupState state = synchronizer.getState();
+        Optional<Map<Stream, StreamCut>> cuts = state.getStreamCutsForCompletedCheckpoint(checkPointId);
+        synchronizer.updateStateUnconditionally(new ClearCheckpointsBefore(checkPointId));
```

Comment:
Look at the implementation of ClearCheckpointsBefore it calls CheckpointState.clearCheckpointsBefore. So the race would be some other unknown update occurs before that method is invoked. That looks harmless to me.

Reply:
I can see that this code is based on `ReaderGroupImpl.completeCheckpoint`.  I feel that I'm missing some piece of the puzzle here. Let me share my insight.

Say that I have two callers C1 and C2 that have requested checkpoints around the same time. Let's call the checkpoints A and B. The checkpoint state in the reader group state is going to have {A, B} as the list of checkpoints. In the case that we complete these checkpoints also around the same time, and we do it unconditionally, then I can have C2 updating after C1, in which case the last state update will have {B} as the list of checkpoints in the checkpoint state, even though B has been completed.

My two questions:
1- Is this scenario possible?
2- Assuming it is possible, is it harmless to have a completed checkpoint in the list?

Reply:
1. No. It's not a 'replace the list with this other list' the command runs like 'locate this item in the ordered list and remove things before it if it is there'. 
So if there is a race and it is already gone the update does nothing. 
2. If we did somehow have a leak It would just waste space. But presumably it would get cleared anyway after the next checkpoint completion.

Reply:
I think the conclusion is that even if the scenario is possible, then it is harmless. But, this observation in point 2:

```
If we did somehow have a leak
```

does not make clear whether the scenario I described is expected or not. It seems to imply that it isn't, but I'd like to have a firm answer so that I know what the expected behavior is. 

Also, point (1) is not really addressing my concern. I understand that the command is not like "replace this list with this other list", and it is as you say. But, it is followed by an unconditional update of the state, which replaces the previous list with the new one generated unconditionally.

Reply:
No. That is not expected. This code does not have a leak.
>But, it is followed by an unconditional update of the state, which replaces the previous list with the new one generated unconditionally.

No it does not do that. Look at the update function. It modifies the list from the one passed which by definition of the StateSynchronizer is the latest one, even inside of an unconditional update